rpcserver.go

```diff
@@ -1079,12 +1079,144 @@ func (r *rpcServer) VerifyProof(ctx context.Context,
 	)
 	valid := err == nil
 
-	// TODO(roasbeef): also show additional final resting anchor
-	// information, etc?
+	decodedProof, err := r.marshalProofFile(ctx, proofFile, 0, false, false)
+	if err != nil {
+		return nil, fmt.Errorf("unable to marshal proof: %w", err)
+	}
+
+	return &taprpc.VerifyProofResponse{
+		Valid:        valid,
+		DecodedProof: decodedProof,
+	}, nil
+}
+
+// DecodeProof attempts to decode a given proof file that claims to be anchored
+// at the specified genesis point.
+func (r *rpcServer) DecodeProof(ctx context.Context,
+	in *taprpc.DecodeProofRequest) (*taprpc.DecodeProofResponse, error) {
+
+	if len(in.RawProof) == 0 {
+		return nil, fmt.Errorf("proof file must be specified")
+	}
+
+	var proofFile proof.File
+	if err := proofFile.Decode(bytes.NewReader(in.RawProof)); err != nil {
+		return nil, fmt.Errorf("unable to decode proof file: %w", err)
+	}
+
+	latestProofIndex := uint32(proofFile.NumProofs() - 1)
+
+	if in.ProofAtDepth > latestProofIndex {
+		return nil, fmt.Errorf("invalid depth %d is greater than latest proof"+
+			" index of %d", in.ProofAtDepth, latestProofIndex)
+	}
+
+	// Default to latest proof.
+	depth := latestProofIndex - in.ProofAtDepth
+
+	decodedProof, err := r.marshalProofFile(
+		ctx, proofFile, depth, in.WithPrevWitnesses, in.WithMetaReveal,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("unable to marshal proof: %w", err)
+	}
+
+	return &taprpc.DecodeProofResponse{
+		DecodedProof: decodedProof,
+	}, nil
+}
+
+// marshalProofFile turns a proof file into an RPC DecodedProof.
+func (r *rpcServer) marshalProofFile(ctx context.Context, proofFile proof.File,
+	depth uint32, withPrevWitnesses, withMetaReveal bool) (*taprpc.DecodedProof,
+	error) {
+
+	decodedProof, err := proofFile.ProofAt(depth)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		finalAsset     = decodedProof.Asset
+		rpcMeta        *taprpc.AssetMeta
+		anchorOutpoint = wire.OutPoint{
+			Hash:  decodedProof.AnchorTx.TxHash(),
+			Index: decodedProof.InclusionProof.OutputIndex,
+		}
+		txMerkleProof  = decodedProof.TxMerkleProof
+		inclusionProof = decodedProof.InclusionProof
+		splitRootProof = decodedProof.SplitRootProof
+	)
+
+	var txMerkleProofBuf bytes.Buffer
+	if err := txMerkleProof.Encode(&txMerkleProofBuf); err != nil {
+		return nil, fmt.Errorf("unable to encode serialized Bitcoin "+
+			"merkle proof: %w", err)
+	}
+
+	var inclusionProofBuf bytes.Buffer
+	if err := inclusionProof.Encode(&inclusionProofBuf); err != nil {
+		return nil, fmt.Errorf("unable to encode inclusion proof: %w",
+			err)
+	}
+
+	var exclusionProofs [][]byte
+	for _, exclusionProof := range decodedProof.ExclusionProofs {
+		var exclusionProofBuf bytes.Buffer
+		if err := exclusionProof.Encode(&exclusionProofBuf); err != nil {
+			return nil, fmt.Errorf("unable to encode exclusion "+
```

Comment:
would be helpful to add the index of the range in this error?

Reply:
Like this?
```
for i, exclusionProof := range decodedProof.ExclusionProofs {
		var exclusionProofBuf bytes.Buffer
		if err := exclusionProof.Encode(&exclusionProofBuf); err != nil {
			return nil, fmt.Errorf("unable to encode exclusion "+
				"proof of index %d: %w", i, err)
		}
		exclusionProofBytes := exclusionProofBuf.Bytes()

		exclusionProofs = append(exclusionProofs, exclusionProofBytes)
	}
	```